src/visualize/VoronoiGeometries.cpp

```diff
@@ -147,20 +147,130 @@ VoronoiGeometry NuTo::Visualize::VoronoiGeometryBrick(int num, Spacing s)
-    geometry.pointCoordinates.push_back(Eigen::Vector2d(0, 1));
-
-    geometry.voronoiCells.push_back({{0, 1, 4, 3}, NuTo::eCellTypes::QUAD});
-    geometry.voronoiCells.push_back({{1, 2, 5, 4}, NuTo::eCellTypes::QUAD});
-    geometry.voronoiCells.push_back({{4, 5, 6, 3}, NuTo::eCellTypes::QUAD});
+    for (size_t i = 0; i < rExisingPoints->size(); ++i)
+        if ((newPoint - (*rExisingPoints)[i]).squaredNorm() < 1.e-4)
+            return i;
+    rExisingPoints->push_back(newPoint);
+    return rExisingPoints->size() - 1;
+}
+
+std::vector<Eigen::Vector2d> TransformedHull(std::vector<Eigen::Vector2d> hullpoints, Eigen::Matrix2d T)
+{
+    std::vector<Eigen::Vector2d> hull;
+    for (Eigen::Vector2d p : hullpoints)
+        hull.push_back(T * p);
+    return hull;
+}
+
+std::vector<VoronoiPoint> TransformedIps(const NuTo::IntegrationTypeBase& integrationType, Eigen::Matrix2d T)
+{
+    std::vector<VoronoiPoint> points;
+    for (int i = 0; i < integrationType.GetNumIntegrationPoints(); ++i)
+    {
+        Eigen::Vector2d ip = T * Eigen::Vector2d(integrationType.GetLocalIntegrationPointCoordinates(i).x(),
+                                                 integrationType.GetLocalIntegrationPointCoordinates(i).y());
+        points.push_back({ip.x(), ip.y()});
+    }
+    return points;
+}
+
+std::vector<VoronoiPoint> ReflectPointsOnHull(std::vector<VoronoiPoint> points, std::vector<Eigen::Vector2d> hull)
+{
+    auto voronoipoints = points;
+    for (size_t iEdge = 0; iEdge < hull.size(); ++iEdge)
+        for (auto point : points)
+        {
+            Eigen::Vector2d a = hull[iEdge];
+            Eigen::Vector2d b = hull[(iEdge + 1) % hull.size()];
+            Eigen::Vector2d n = (b - a).normalized();
+            Eigen::Vector2d p(point.x(), point.y());
 
+            Eigen::Vector2d pNew = a + 2 * n * n.dot(p - a) - (p - a);
+            voronoipoints.push_back({pNew.x(), pNew.y()});
+        }
+    return voronoipoints;
+}
+
+/* Voronoi diagrams may have infinite edges like:
+ *        |
+ *     x / \ x
+ *      / x \
+ *
+ * Thus, we have to somehow consider the hull. We simply do that by reflecting all points on that hull
+ * [ReflectPointsOnHull] and do voronoi on the original _and_ the reflected points.
+ *
+ * [Performace is not needed here, since the geometry is calculated once and stored.]
+ *
+ * x   x  / \  x   x
+ *       / x \
+ *   x  /     \ x
+ *     / x   x \
+ *    /_________\
+ *
+ *       x   x
+ *
+ *         x
+ *
+ * This will ensure that the voronoi diagramm has an edge on the hull.
+ *
+ * boost::polygon::voronoi now creates _cells_. We only evaluate cells that are associated with the origninal points and
+ * transform them to a NuTo::Visualize::VoronoiGeometry. [AddUnique] will make sure that we add points at the same
+ * coordinate only once.
+ */
+NuTo::Visualize::VoronoiGeometry TransformedVoronoi(const NuTo::IntegrationTypeBase& integrationType,
```

Comment:
Personally, I would split that into two functions. One `boost::polygon::voronoi_diagram<double> CreateVoronoiDiagram(...)` with the `TransformedHull` stuff etc, and one `ConvertToVisualizeCells` that takes a boost voronoi diagram and returns a `VoronoiGeometry`.

Reply:
That sounds good. However, `ConvertToVisualizeCells` also involves the back transformation. But I could do the transformation afterwards...

~~~
auto voronoi = CreateVoronoiDiagramm(hull, integration, T);
auto geometry = ConvertToVisualizeCells(voronoi);
TransformBack(&geometry, T);
~~~
?